Add WorkerClient for WebCore to access WebKit from worker threads, and use it to create a RemoteRenderingBackendProxy per-thread

mattwoodrow opened the pull request:
#### 280e6cdf0e10c5649b60dade69a0e43bd9108037
<pre>
Add WorkerClient for WebCore to access WebKit from worker threads, and use it to create a RemoteRenderingBackendProxy per-thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=244828">https://bugs.webkit.org/show_bug.cgi?id=244828</a>

Reviewed by Kimmo Kinnunen.

Creates a new interface class WorkerClient (which implements GraphicsClient), for accessing WebKit graphics APIs from a worker thread. Adds an implementation
of this in WebWorkerClient.

Allocates an instance of (Web)WorkerClient for every dedicated and shared worker, and makes it available via WorkerGlobalScope.

WebWorkerClient creates a dedicated GPUProcessConnection and RemoteRenderingBackendProx for each worker thread/instance, and uses this to create
RemoteImageBufferProxys that are safe to use on that worker thread.

* Source/WebCore/Headers.cmake:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::createWorkerClient):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::createWorkerClient):
* Source/WebCore/page/WorkerClient.h: Added.
* Source/WebCore/workers/DedicatedWorkerGlobalScope.cpp:
(WebCore::DedicatedWorkerGlobalScope::create):
(WebCore::DedicatedWorkerGlobalScope::DedicatedWorkerGlobalScope):
* Source/WebCore/workers/DedicatedWorkerGlobalScope.h:
* Source/WebCore/workers/DedicatedWorkerThread.cpp:
(WebCore::DedicatedWorkerThread::createWorkerGlobalScope):
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::WorkerGlobalScope):
* Source/WebCore/workers/WorkerGlobalScope.h:
(WebCore::WorkerGlobalScope::workerClient):
* Source/WebCore/workers/WorkerMessagingProxy.cpp:
(WebCore::WorkerMessagingProxy::startWorkerGlobalScope):
* Source/WebCore/workers/WorkerThread.h:
(WebCore::WorkerThread::setWorkerClient):
(WebCore::WorkerThread::workerClient):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::ServiceWorkerGlobalScope):
* Source/WebCore/workers/shared/SharedWorkerGlobalScope.cpp:
(WebCore::SharedWorkerGlobalScope::SharedWorkerGlobalScope):
* Source/WebCore/workers/shared/SharedWorkerGlobalScope.h:
* Source/WebCore/workers/shared/context/SharedWorkerThread.cpp:
(WebCore::SharedWorkerThread::createWorkerGlobalScope):
* Source/WebCore/workers/shared/context/SharedWorkerThreadProxy.cpp:
(WebCore::SharedWorkerThreadProxy::SharedWorkerThreadProxy):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWorkerClient):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp: Added.
(WebKit::WebWorkerClient::WebWorkerClient):
(WebKit::WebWorkerClient::ensureRenderingBackend const):
(WebKit::WebWorkerClient::clone):
(WebKit::WebWorkerClient::displayID const):
(WebKit::WebWorkerClient::createImageBuffer const):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.h: Added.

Canonical link: <a href="https://commits.webkit.org/257507@main">https://commits.webkit.org/257507@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7088983bee320a698f700fc39a7557388185a7eb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98426 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7631 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31555 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107852 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168118 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102363 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8144 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85011 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90973 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104510 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104090 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6191 "Found 30 new test failures: editing/selection/ios/change-selection-by-tapping.html, fast/canvas/image-pattern-rotate.html, fast/css/css-typed-om/style-property-map-set-CSSMathSum-value.html, fast/text/firstline/001.html, http/tests/app-privacy-report/app-attribution-beacon-isappinitiated.html, http/tests/app-privacy-report/app-attribution-beacon-isnotappinitiated.html, http/tests/app-privacy-report/user-attribution-cors-preflight-redirect.html, http/tests/blink/sendbeacon/beacon-cookie.html, http/tests/cache/disk-cache/disk-cache-remove-several-pending-writes.html, imported/w3c/web-platform-tests/css/css-flexbox/flexbox_first-letter.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89746 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33181 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87996 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21105 "Found 30 new test failures: accessibility/aria-keyshortcuts.html, animations/additive-transform-animations.html, compositing/no-compositing-when-fulll-screen-is-present.html, fast/css/css-typed-om/style-property-map-set-CSSMathSum-value.html, fast/text/text-edge-property-parsing.html, http/tests/appcache/fail-on-update-2.html, http/tests/appcache/remove-cache.html, http/tests/media/fairplay/fps-hls-key-rotation.html, http/tests/media/fairplay/fps-mse-unmuxed-key-rotation.html, imported/w3c/web-platform-tests/content-security-policy/reporting-api/reporting-api-works-on-frame-src.https.sub.html ... (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76127 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1575 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22633 "Found 30 new test failures: accessibility/aria-keyshortcuts.html, compositing/no-compositing-when-fulll-screen-is-present.html, fast/css/css-typed-om/style-property-map-set-CSSMathSum-value.html, fast/rendering/render-style-null-optgroup-crash.html, fast/text/text-edge-property-parsing.html, http/tests/xmlhttprequest/loadstart-event-init.html, http/wpt/cache-storage/cache-quota.any.html, http/wpt/fetch/response-opaque-clone.html, imported/w3c/web-platform-tests/content-security-policy/reporting/report-uri-scheme-relative.html, imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml.html ... (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1498 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45123 "Found 30 new test failures: accessibility/aria-keyshortcuts.html, animations/stop-animation-on-suspend.html, compositing/no-compositing-when-fulll-screen-is-present.html, fast/css/css-typed-om/style-property-map-set-CSSMathSum-value.html, fast/rendering/render-style-null-optgroup-crash.html, fast/text/text-edge-property-parsing.html, http/wpt/service-workers/fetch-service-worker-preload-download.https.html, http/wpt/service-workers/fetch-service-worker-preload.https.html, imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml.html, imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/masonry-grid-template-columns-computed-withcontent.html ... (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5160 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6421 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42048 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2873 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -92,15 +95,19 @@ RefPtr<ImageBuffer> ImageBitmap::createImageBuffer(ScriptExecutionContext& scrip
         }
+    } else if (scriptExecutionContext.isWorkerGlobalScope()) {
+        client = downcast<WorkerGlobalScope>(scriptExecutionContext).getWorkerClient();
     }
```
I think curly braces are not needed here.

shallawa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -92,15 +95,19 @@ RefPtr<ImageBuffer> ImageBitmap::createImageBuffer(ScriptExecutionContext& scrip
+    if (client)
+        return ImageBuffer::create(size, RenderingPurpose::Canvas, resolutionScale, *imageBufferColorSpace, PixelFormat::BGRA8, bufferOptions, { client });
+
     return ImageBuffer::create(size, RenderingPurpose::Unspecified, resolutionScale, *imageBufferColorSpace, PixelFormat::BGRA8, bufferOptions);
```
Why do not we pass `RenderingPurpose::Canvas` and `client` even if it is nullptr to `ImageBuffer::create()`? `ImageBuffer::create()` has its own fallback and it is used by all the callers.

shallawa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -113,10 +120,29 @@ void ImageBitmap::createPromise(ScriptExecutionContext& scriptExecutionContext,
+RefPtr<ImageBuffer> ImageBitmap::createImageBuffer(ScriptExecutionContext& scriptExecutionContext, const FloatSize& size, DestinationColorSpace colorSpace, float resolutionScale)
+{
+    return createImageBuffer(scriptExecutionContext, size, bufferRenderingMode, colorSpace, resolutionScale);
+}
```
There is a version of `ImageBitmap::createImageBuffer()` which is similar to this but it an `IntSize` and does not take a `resolutionScale`. Should we add a new argument to it 'float resolutionScale = 1' instead of creating a new one?

shallawa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -113,10 +120,29 @@ void ImageBitmap::createPromise(ScriptExecutionContext& scriptExecutionContext,
+        // FIXME: Currently remote-backed ImageBuffers can't be transferred to a different thread,
+        // so we have to read the contents back to a local ImageBuffer and transfer that instead.
+        // Bug 244830 will add support for direct transfers.
+        if (backing && backing->buffer() && backing->buffer()->isRemote()) {
```
Maybe an early return makes code clearer:
```
if (!backing || !...)
    return backing;
```
Also I think adding isRemote() is not the correct approach.

shallawa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -113,10 +120,29 @@ void ImageBitmap::createPromise(ScriptExecutionContext& scriptExecutionContext,
+        // Bug 244830 will add support for direct transfers.
+        if (backing && backing->buffer() && backing->buffer()->isRemote()) {
+            auto source = backing->buffer();
+            auto copyBuffer = ImageBuffer::create(source->logicalSize(), RenderingPurpose::Unspecified, source->resolutionScale(), source->colorSpace(), source->pixelFormat());
```
I do not understand this code and the FIXME comment above. When is` ImageBitmap::detachBitmaps()` called? Is creating a local unaccelerated ImageBuffer what we want in this case?

mattwoodrow replied:
`detachBitmaps` gets called when JS uses `postMessage()` to transfer an `ImageBitmap` to/from a worker.

I wrote a longer explanation below, the answer is no, we don't really want that, but the right fix is a large patch in its own right (and is coming!).

shallawa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -897,10 +923,6 @@ ImageBitmap::ImageBitmap(std::optional<ImageBitmapBacking>&& backingStore)
-        return;
-    if (auto imageBuffer = takeImageBuffer())
-        callOnMainThread([imageBuffer = WTFMove(imageBuffer)] { });
 }
```
Can't we remove this destructor altogether?

shallawa commented on Source/WebCore/html/OffscreenCanvas.cpp:
```diff
@@ -295,7 +295,7 @@ ExceptionOr<RefPtr<ImageBitmap>> OffscreenCanvas::transferToImageBitmap()
 
         if (!m_hasCreatedImageBuffer) {
-            auto buffer = ImageBitmap::createImageBuffer(*canvasBaseScriptExecutionContext(), size(), RenderingMode::Unaccelerated, m_context->colorSpace());
+            auto buffer = ImageBitmap::createImageBuffer(*canvasBaseScriptExecutionContext(), size(), m_context->colorSpace());
```
Why did we use `RenderingMode::Unaccelerated` before?

shallawa commented on Source/WebCore/page/Chrome.h:
```diff
@@ -25,7 +25,9 @@
 #include "DisabledAdaptations.h"
 #include "FocusDirection.h"
 #include "HostWindow.h"
+#include "ImageBufferBackend.h"
```
Why do we need to include this header file here?

shallawa commented on Source/WebCore/platform/GraphicsClient.h:
```diff
@@ -0,0 +1,50 @@
+
+#pragma once
+
+#include "PlatformScreen.h"
```
Instead of including the whole file, I think you can just say
```
using PlatformDisplayID = uint32_t;

```
Searching for this string in the WebKit gives 12 files which have the same string. I think we need to move this definition to a separate file.

shallawa commented on Source/WebCore/platform/HostWindow.h:
```diff
@@ -80,7 +73,6 @@ class HostWindow {
 
     virtual void setCursorHiddenUntilMouseMoves(bool) = 0;
 
-    virtual PlatformDisplayID displayID() const = 0;
```
I am not the right person to say if this is correct or not. I think @hortont424 knows better. But I think the `PlatformDisplayID` should be associated with Window, Screen or Page but not with a display element. So moving this method to `GraphicsClient` does not look right to me.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -161,6 +161,7 @@ class ImageBuffer : public ThreadSafeRefCounted<ImageBuffer, WTF::DestructionThr
     RenderingPurpose renderingPurpose() const { return m_parameters.purpose; }
     PixelFormat pixelFormat() const { return m_parameters.pixelFormat; }
     const ImageBufferBackend::Parameters& parameters() const { return m_parameters; }
+    virtual bool isRemote() const { return false; }
```
We have been trying hard to not add this specific function. I think @weinig will not be happy with this change.

From the beginning we did not want to have any code in WebCore which makes decision based on whether we renders in WebProcess or we render in GPUProcess. All these decisions have to be in WebKit.

So please move all the GPUP decisions to WebKit.

mattwoodrow replied:
I am struggling to balance between making this change small enough to be reviewable, but also making sure it works.

The existing state is that remote ImageBuffers only really work on the main-thread, but we let JS postMessage them into workers, and we just use `DestructionThread::Main` to make sure they get sent back if they get destroyed. Since OffscreenCanvas is disabled, there's no Web APIs that can use them on a worker, so they can exist on a worker without breaking anything.

This patch makes workers have their own remote rendering backend proxy, so that remote ImageBuffers created on a worker are now actually usable. Unfortunately, we now need to do something different when we copy an ImageBuffer between threads, so this patch forces read back into a local unaccelerated buffer.

I added `isRemote` as a temporary hack to know when to do this, to avoid regressing performance for Gtk (which does have offscreen canvas enabled).

Bug 244830 is going to add proper support for transferring ImageBuffers between threads (and changing them to use the appropriate remote rendering backend proxy), and I can remove `isRemote` again as part of that.

mattwoodrow replied:
The transfer changes are here: https://github.com/mattwoodrow/WebKit/commit/ee88514e6c393d65f8b47512aaf53b6582cd02e38c

Still needs some cleanup, but we won't need `isRemote` with those.

shallawa commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -362,14 +362,14 @@ static IntSize computeMaximumSurfaceSize()
-        if (hostWindow && hostWindow->displayID())
-            displayMask = displayMaskForDisplay(hostWindow->displayID());
+        if (client && client->displayID())
+            displayMask = displayMaskForDisplay(client->displayID());
```
Okay now I see why you needed the `displayID` in the `GraphicsClient`. But I still do not think that every `GraphicsClient` should have a `PlatformDisplayID`. I think only the root `HostWindow` can have `PlatformDisplayID`.

shallawa commented on Source/WebCore/workers/DedicatedWorkerThread.cpp:
```diff
@@ -55,6 +55,7 @@ Ref<WorkerGlobalScope> DedicatedWorkerThread::createWorkerGlobalScope(const Work
         scope->setActiveServiceWorker(ServiceWorker::getOrCreate(scope.get(), ServiceWorkerData { *params.serviceWorkerData }));
     scope->updateServiceWorkerClientData();
 #endif
+    scope->setWorkerClient(WTFMove(m_workerClient));
```
Where `m_workerClient` is created? And why it is not passed to the `DedicatedWorkerGlobalScope::create()` instead of setting it later?

mattwoodrow replied:
It's created in `WorkerMessagingProxy::startWorkerGlobalScope`.

shallawa commented on Source/WebCore/platform/GraphicsClient.h:
```diff
@@ -0,0 +1,50 @@
+    GraphicsClient() = default;
+    virtual ~GraphicsClient() = default;
+
+    virtual PlatformDisplayID displayID() const = 0;
```
My understanding about `PlatformDisplayID` is it is a screen ID. So why `GraphicsClient`, which can be an `ImageBitmap`, has a `PlatformDisplayID`?

mattwoodrow replied:
`ImageBitmap` is not a `GraphicsClient`, only `HostWindow` and `WorkerClient`.So only a real window, or a worker created on behalf of a window should have the display ID.

The `WorkerClient` is a subsidiary of the window, but needs to have it's own copy of the displayID, since it can't access the main window object from another thread.

shallawa commented on Source/WebCore/workers/WorkerGlobalScope.h:
```diff
@@ -167,8 +168,10 @@ class WorkerGlobalScope : public Supplementable<WorkerGlobalScope>, public Base6
 
     ClientOrigin clientOrigin() const { return { topOrigin().data(), securityOrigin()->data() }; }
 
+    WorkerClient* getWorkerClient() { return m_workerClient.get(); }
```
We usually do not use the verb "get" with the getter functions. I think this should be just `workerClient()`.

shallawa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -113,10 +116,18 @@ void ImageBitmap::createPromise(ScriptExecutionContext& scriptExecutionContext,
-        return bitmap->takeImageBitmapBacking();
+    return WTF::map(WTFMove(bitmaps), [](auto&& bitmap) -> std::optional<ImageBitmapBacking> {
+        std::optional<ImageBitmapBacking> backing = bitmap->takeImageBitmapBacking();
+        if (auto copyBuffer = ImageBuffer::sinkIntoBufferForDifferentThread(backing->takeImageBuffer()))
```
Maybe we can add a new function `ImageBitmapBacking::takeImageBufferForDifferentThread()`. So this line can be:

```
if (auto buffer = backing->takeImageBufferForDifferentThread())
```

shallawa commented on Source/WebCore/html/OffscreenCanvas.cpp:
```diff
@@ -441,7 +441,9 @@ std::unique_ptr<DetachedOffscreenCanvas> OffscreenCanvas::detach()
-    auto detached = makeUnique<DetachedOffscreenCanvas>(takeImageBuffer(), size(), originClean());
+    auto buffer = ImageBuffer::sinkIntoBufferForDifferentThread(takeImageBuffer());
+
+    auto detached = makeUnique<DetachedOffscreenCanvas>(WTFMove(buffer), size(), originClean());
```
`auto detached = makeUnique<DetachedOffscreenCanvas>(takeImageBufferForDifferentThread(),...);`?

shallawa commented on Source/WebCore/html/OffscreenCanvas.cpp:
```diff
@@ -487,7 +489,7 @@ void OffscreenCanvas::commitToPlaceholderCanvas()
     Locker locker { m_placeholderData->bufferLock };
     bool shouldPushBuffer = !m_placeholderData->pendingCommitBuffer;
-    m_placeholderData->pendingCommitBuffer = imageBuffer->clone();
+    m_placeholderData->pendingCommitBuffer = ImageBuffer::sinkIntoBufferForDifferentThread(imageBuffer->clone());
```
Can we add `ImageBuffer::cloneForDifferentThread()` instead of having to make an extra clone?

shallawa commented on Source/WebCore/platform/GraphicsClient.h:
```diff
@@ -0,0 +1,50 @@
+enum class RenderingMode : bool;
+enum class RenderingPurpose : uint8_t;
+
+class GraphicsClient {
```
What does exactly this class represent? For me it does not deal with real graphics drawing or composting. I think it is more or less an `ImageBufferFactory`. Please consider renaming.

mattwoodrow replied:
It's the common functionality between HostWindow/ChromeClient, and WorkerClient, which is all graphics related.

Right now it just has a function to create ImageBuffers, but https://github.com/WebKit/WebKit/pull/5842 adds WebGL too (and I assume in the future we'll want WebGPU).

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -227,13 +227,28 @@ RefPtr<NativeImage> ImageBuffer::copyNativeImageForDrawing(BackingStoreCopy copy
 
 RefPtr<NativeImage> ImageBuffer::sinkIntoNativeImage()
 {
+
```
extra line.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -49,11 +49,11 @@ enum class ImageBufferOptions : uint8_t {
     UseDisplayList  = 1 << 1
 };
 
-class ImageBuffer : public ThreadSafeRefCounted<ImageBuffer, WTF::DestructionThread::Main>, public CanMakeWeakPtr<ImageBuffer> {
```
This is a nice change. But I think we need to review the changes we did, because of having to destroy the ImageBuffer on the main thread, especially the ones on RemoteRenderingBackend.

mattwoodrow replied:
The changes for ImageBuffer to use DestructionThread::Main were to handle RemoteRenderingBackendProxy (in the WebContent process). Previously we allowed these to be sent to a worker, and if they got destroyed there they attempted to use IPC from the wrong thread.

This patches solves this by making sure we don't copy the RRBP between thread, and remote proxy buffers now strongly belong to their thread, so DestructionThread::Main is incorrect.

In the GPUP, DestructionThread::Main wasn't necessary (as far as I can tell), and potentially causes issues with WeakPtrs (which aren't threadsafe), since we hold weak refs and send them to be destroyed on a different thread.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
```diff
@@ -93,7 +93,7 @@ std::unique_ptr<ImageBufferIOSurfaceBackend> ImageBufferIOSurfaceBackend::create
         return nullptr;
 
-    RetainPtr<CGContextRef> cgContext = surface->ensurePlatformContext(creationContext.hostWindow);
+    RetainPtr<CGContextRef> cgContext = surface->ensurePlatformContext(creationContext.graphicsClient);
```
`ensurePlatformContext()` needs only the `displayID` from `graphicsClient`. Can we change it to a `displayID` instead? `displayID` with value equal to zero is invalid.

shallawa commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -350,7 +350,7 @@ static IntSize computeMaximumSurfaceSize()
 }
 
-CGContextRef IOSurface::ensurePlatformContext(const HostWindow* hostWindow)
+CGContextRef IOSurface::ensurePlatformContext(const GraphicsClient* client)
```
This can be 

```
CGContextRef IOSurface::ensurePlatformContext(DisplayID displayID)
```

shallawa commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -362,14 +362,14 @@ static IntSize computeMaximumSurfaceSize()
     if (auto displayMask = primaryOpenGLDisplayMask()) {
-        if (hostWindow && hostWindow->displayID())
-            displayMask = displayMaskForDisplay(hostWindow->displayID());
+        if (client && client->displayID())
```
This can be 

```
if (displayID)
```

shallawa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -92,16 +95,16 @@ RefPtr<ImageBuffer> ImageBitmap::createImageBuffer(ScriptExecutionContext& scrip
         }
-    }
+    } else if (scriptExecutionContext.isWorkerGlobalScope())
+        client = downcast<WorkerGlobalScope>(scriptExecutionContext).getWorkerClient();
```
`workerClient()`?

shallawa commented on Source/WebCore/workers/WorkerMessagingProxy.cpp:
```diff
@@ -142,8 +143,17 @@ void WorkerMessagingProxy::startWorkerGlobalScope(const URL& scriptURL, PAL::Ses
+        if (parentWorkerGlobalScope->thread().workerClient())
+            thread->setWorkerClient(parentWorkerGlobalScope->thread().workerClient()->clone(thread.get()));
+    } else if (is<Document>(m_scriptExecutionContext.get())) {
+        Document& document = downcast<Document>(*m_scriptExecutionContext);
```
auto&?

shallawa commented on Source/WebCore/workers/shared/context/SharedWorkerThreadProxy.cpp:
```diff
@@ -108,6 +110,9 @@ SharedWorkerThreadProxy::SharedWorkerThreadProxy(UniqueRef<Page>&& page, SharedW
         addedListener = true;
     }
+
+    if (auto workerClient = m_page->chrome().createWorkerClient(thread()))
+        thread().setWorkerClient(WTFMove(workerClient));
```
It is a little strange that we pass thread() to create the worker client. When it is created, we call thread().setWorkerClient(...). Can't everything be done in WorkerThread? And here we can say:

```
thread().createWorkerClient(m_page->chrome());
```

mattwoodrow replied:
It's a bit tougher for the nested worker case (for dedicated workers), where we won't have a `chrome()` object to pass in.

I could create a second overload of `createWorkerClient` for that I guess?

shallawa commented on Source/WebKit/GPUProcess/GPUProcess.cpp:
```diff
@@ -140,15 +140,33 @@ void GPUProcess::createGPUConnectionToWebProcess(WebCore::ProcessIdentifier iden
+        ASSERT(!connections.primary);
+        connections.primary = newConnection.ptr();
+    } else
+        connections.secondary.append(newConnection);
```
I think `secondary` is unclear name especially it is a Vector. Can't we use a plural name `workers` for example?

shallawa commented on Source/WebKit/GPUProcess/GPUProcess.cpp:
```diff
@@ -348,7 +374,18 @@ void GPUProcess::resume()
+    if (iter == m_webProcessConnections.end())
+        return nullptr;
+    return &iter->value.secondary;
 }
```
This function is never called in this patch. Should it be removed?

shallawa commented on Source/WebKit/GPUProcess/GPUProcess.h:
```diff
@@ -188,7 +190,11 @@ class GPUProcess : public AuxiliaryProcess, public ThreadSafeRefCounted<GPUProce
+        RefPtr<GPUConnectionToWebProcess> primary;
+        Vector<Ref<GPUConnectionToWebProcess>> secondary;
+    };
+    HashMap<WebCore::ProcessIdentifier, WebProcessConnections> m_webProcessConnections;
```
This structure does not seem very elegant. In most of the code above you loop through `m_webProcessConnections` and every loop you check the primary connection and then you loop through the secondary connections. I think having the value of the `m_webProcessConnections` be a `Vector<RefPtr<GPUConnectionToWebProcess>>` and dedicating the first place for primary connection will avoid repeating this pattern.

mattwoodrow replied:
I had that originally, my worry was that it was hard to guarantee (statically, at least) that the primary connection was always created first and deleted last. We don't ever want a state where a non-primary connection is in slot 0.

shallawa commented on Source/WebKit/GPUProcess/GPUProcess.cpp:
```diff
@@ -548,20 +593,24 @@ void GPUProcess::webProcessConnectionCountForTesting(CompletionHandler<void(uint
+            connection.primary->processIsStartingToCaptureAudio(process);
+        for (auto& secondary : connection.secondary)
+            secondary->processIsStartingToCaptureAudio(process);
+    }
```
This pattern is repeated multiple times. I think we can do a better job by having either an enumeration function or a new structure for the values in m_webProcessConnections.

```
enumerate(connections, [&](GPUConnectionToWebProcess& connection) {
    connection.processIsStartingToCaptureAudio(process);
};
```

shallawa commented on Source/WebKit/Shared/GPUProcessConnectionParameters.h:
```diff
@@ -37,6 +37,7 @@ struct GPUProcessConnectionParameters {
     WebCore::ProcessIdentity webProcessIdentity;
     Vector<String> overrideLanguages;
     bool isLockdownModeEnabled { false };
+    bool isForMainThread { true };
```
Can we stick with one name `forMainThread` (and `isForMainThread`) or `primary` (and `isPrimary`) which is defined in `WebProcessConnections`?

shallawa commented on Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
```diff
@@ -109,36 +109,39 @@ static void languagesChanged(void* context)
     parameters.webProcessIdentity = ProcessIdentity { ProcessIdentity::CurrentProcess };
     parameters.overrideLanguages = userPreferredLanguagesOverride();
 #endif
+    parameters.isForMainThread = forMainThread;
```
Can we either include `is` in both or remove it from both: the member and the argument?

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -227,13 +227,28 @@ RefPtr<NativeImage> ImageBuffer::copyNativeImageForDrawing(BackingStoreCopy copy
+    return buffer->sinkIntoBufferForDifferentThread();
+}
+
+RefPtr<ImageBuffer> ImageBuffer::sinkIntoBufferForDifferentThread()
```
This name is confusing. It does not sink the current ImageBuffer. It just make a new RefPtr of it.

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
```diff
@@ -194,6 +202,17 @@ RefPtr<NativeImage> RemoteImageBufferProxy::sinkIntoNativeImage()
     return copyNativeImage();
 }
 
+RefPtr<ImageBuffer> RemoteImageBufferProxy::sinkIntoBufferForDifferentThread()
```
This name is confusing. It is just returns a local copy of the remote ImageBuffer with RenderingPurpose::Unspecified. It does not touch the backend of this remote ImageBuffer. 

I would suggest renaming this function to `cloneLocal()` or `copyLocal()` or `cloneForDifferentThread()` or something like that:

```
RefPtr<ImageBuffer> RemoteImageBufferProxy::cloneForDifferentThread() const
```

And make ImageBuffer::sinkIntoBufferForDifferentThread(RefPtr<ImageBuffer>) calls it with the new name. I think this will make sense because the static method will sink the ImageBuffer to a new local (or existing local) ImageBuffer. And the virtual method will just do the local copy.

mattwoodrow replied:
Does that mean that the base class impl of `cloneForDifferentThread` would still just `return this`? Is that not also confusing?

The intent is that the static function consumes the existing buffer, and returns a buffer that is safe to pass to a new thread. That can be either by returning the same object back again (if it were already safe), or by creating one that is and copying the contents across.

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -48,15 +48,17 @@ using namespace WebCore;
-        webPage.webPageProxyIdentifier(),
-        webPage.identifier()
-    }
+std::unique_ptr<RemoteRenderingBackendProxy> RemoteRenderingBackendProxy::create(const RemoteRenderingBackendCreationParameters& parameters, SerialFunctionDispatcher* dispatcher)
```
Why `dispatcher` is a pointer? Should not it be a reference?

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -48,15 +48,17 @@ using namespace WebCore;
+
+RemoteRenderingBackendProxy::RemoteRenderingBackendProxy(const RemoteRenderingBackendCreationParameters& parameters, SerialFunctionDispatcher* dispatcher)
+    : m_parameters(parameters)
+    , m_dispatcher(dispatcher ? *dispatcher : RunLoop::main())
```
Can `dispatcher` be null? I think it is either RunLoop::main() or valid `SerialFunctionDispatcher`.

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -74,13 +76,16 @@ RemoteRenderingBackendProxy::~RemoteRenderingBackendProxy()
+        if (&m_dispatcher == &RunLoop::main())
+            m_gpuProcessConnection = &WebProcess::singleton().ensureGPUProcessConnection();
+        else
+            m_gpuProcessConnection = WebProcess::singleton().createGPUProcessConnection(&m_dispatcher);
```
The gpu connection can be lost and restored. Do we need to create it every time for the workers? Or should we just ensure it is created for the current `m_dispatcher` like what we do for the main thread in `ensureGPUProcessConnection()`?

shallawa commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1296,6 +1296,11 @@ WebLoaderStrategy& WebProcess::webLoaderStrategy()
 
+RefPtr<GPUProcessConnection> WebProcess::createGPUProcessConnection(SerialFunctionDispatcher* dispatcher)
+{
+    return GPUProcessConnection::create(*singleton().parentProcessConnection(), dispatcher);
```
Can dispatcher be null?

shallawa commented on Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp:
```diff
@@ -0,0 +1,98 @@
+
+#if ENABLE(GPU_PROCESS)
+WebWorkerClient::WebWorkerClient(RemoteRenderingBackendCreationParameters& creationParameters, SerialFunctionDispatcher& dispatcher, WebCore::PlatformDisplayID& displayID)
+    : m_dispatcher(dispatcher)
+    , m_creationParameters(creationParameters)
+    , m_displayID(displayID)
```
I think it is more readable when we arrange the constructor arguments in the same order we initialize them which is the order of how they are defined in the class definition.

shallawa commented on Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.h:
```diff
@@ -0,0 +1,74 @@
+class WebPage;
+class RemoteRenderingBackendProxy;
+
+class WebWorkerClient : public WebCore::WorkerClient {
```
There are many `#if ENABLE(GPU_PROCESS)` in this header file and in the source file. Should not the GPUP case be moved to a new class named `RemoteWebWorkerClient ` which inherits from `WebWorkerClient `? `WebChromeClient::createWorkerClient()` will make the decision from the beginning so we do not have to deal with two different code paths in the same file.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -227,13 +227,28 @@ RefPtr<NativeImage> ImageBuffer::copyNativeImageForDrawing(BackingStoreCopy copy
+
+RefPtr<ImageBuffer> ImageBuffer::sinkIntoBufferForDifferentThread()
+{
+    ASSERT(hasOneRef());
```
Why is this always true? And what is the point in asserting it hasOneRef() and then create a new RefPtr of the ImageBuffer which will increment the ref count?

mattwoodrow replied:
As above, the naming might not be great here.

It asserts that there's only one ref, because it's a private virtual and the only caller is the static which also asserts that it owns the only ref.

The default behaviour is that if the static owns the only ref to the ImageBuffer, then we can allow it to be passed to a new thread as-is since there won't be any other cross-thread references.

shallawa commented on Source/WebKit/GPUProcess/GPUProcess.cpp:
```diff
@@ -548,20 +593,24 @@ void GPUProcess::webProcessConnectionCountForTesting(CompletionHandler<void(uint
 {
-    for (auto& connection : m_webProcessConnections.values())
-        connection->processIsStartingToCaptureAudio(process);
+    for (auto& connection : m_webProcessConnections.values()) {
```
I think the variable should be named `connections` since its type is `WebProcessConnections`.

shallawa commented on Source/WebKit/GPUProcess/GPUProcess.cpp:
```diff
@@ -548,20 +593,24 @@ void GPUProcess::webProcessConnectionCountForTesting(CompletionHandler<void(uint
+        if (connection.primary)
+            connection.primary->processIsStartingToCaptureAudio(process);
+        for (auto& secondary : connection.secondary)
+            secondary->processIsStartingToCaptureAudio(process);
```
`for (auto&  connection : connections.secondary)`

kkinnunen-apple reviewed (approved):
To solve the most of the ifdefing problems, you can obtain the data needed by RemoteRenderingBackendCreationParameters expanded to the constructor args.

mattwoodrow committed:
Add WorkerClient for WebCore to access WebKit from worker threads, and use it to create a RemoteRenderingBackendProxy per-thread
https://bugs.webkit.org/show_bug.cgi?id=244828

Reviewed by Kimmo Kinnunen.

Creates a new interface class WorkerClient (which implements GraphicsClient), for accessing WebKit graphics APIs from a worker thread. Adds an implementation
of this in WebWorkerClient.

Allocates an instance of (Web)WorkerClient for every dedicated and shared worker, and makes it available via WorkerGlobalScope.

WebWorkerClient creates a dedicated GPUProcessConnection and RemoteRenderingBackendProx for each worker thread/instance, and uses this to create
RemoteImageBufferProxys that are safe to use on that worker thread.

* Source/WebCore/Headers.cmake:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::createWorkerClient):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::createWorkerClient):
* Source/WebCore/page/WorkerClient.h: Added.
* Source/WebCore/workers/DedicatedWorkerGlobalScope.cpp:
(WebCore::DedicatedWorkerGlobalScope::create):
(WebCore::DedicatedWorkerGlobalScope::DedicatedWorkerGlobalScope):
* Source/WebCore/workers/DedicatedWorkerGlobalScope.h:
* Source/WebCore/workers/DedicatedWorkerThread.cpp:
(WebCore::DedicatedWorkerThread::createWorkerGlobalScope):
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::WorkerGlobalScope):
* Source/WebCore/workers/WorkerGlobalScope.h:
(WebCore::WorkerGlobalScope::workerClient):
* Source/WebCore/workers/WorkerMessagingProxy.cpp:
(WebCore::WorkerMessagingProxy::startWorkerGlobalScope):
* Source/WebCore/workers/WorkerThread.h:
(WebCore::WorkerThread::setWorkerClient):
(WebCore::WorkerThread::workerClient):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::ServiceWorkerGlobalScope):
* Source/WebCore/workers/shared/SharedWorkerGlobalScope.cpp:
(WebCore::SharedWorkerGlobalScope::SharedWorkerGlobalScope):
* Source/WebCore/workers/shared/SharedWorkerGlobalScope.h:
* Source/WebCore/workers/shared/context/SharedWorkerThread.cpp:
(WebCore::SharedWorkerThread::createWorkerGlobalScope):
* Source/WebCore/workers/shared/context/SharedWorkerThreadProxy.cpp:
(WebCore::SharedWorkerThreadProxy::SharedWorkerThreadProxy):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWorkerClient):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp: Added.
(WebKit::WebWorkerClient::WebWorkerClient):
(WebKit::WebWorkerClient::ensureRenderingBackend const):
(WebKit::WebWorkerClient::clone):
(WebKit::WebWorkerClient::displayID const):
(WebKit::WebWorkerClient::createImageBuffer const):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.h: Added.

Canonical link: https://commits.webkit.org/257507@main
